src/log_rate.c

```diff
@@ -0,0 +1,210 @@
+#define _GNU_SOURCE
+#include <stdio.h>
+#include <unistd.h>
+#include <string.h>
+#include <time.h>
+#include <stdlib.h>
+#include <errno.h>
+
+#include "utils.h"
+#include "log_rate.h"
+#include "ctr_logging.h"
+
+#define IO_BUF_SIZE      65536
+#define SECS_PER_PERIOD  1
+#define BILLION          1000000000
+
+static const struct timespec secs_per_period = {SECS_PER_PERIOD, 0};
+static size_t bytes_written_this_period = 0;
+static struct timespec start_of_this_period;
+static log_policy_t log_policy;
+static size_t bytes_per_period;
+static bool dropping = false;
+static struct timespec drop_until;
+
+static int64_t add_timespecs_nano(const struct timespec* first, const struct timespec* second);
+static struct timespec add_timespecs(const struct timespec* first, const struct timespec* second);
+static struct timespec subtract_timespecs(const struct timespec* first, const struct timespec* second);
+static int64_t subtract_timespecs_nano(const struct timespec* first, const struct timespec* second);
+static void write_io_bufs(stdpipe_t pipe, char* buf, ssize_t count);
+static void sleep_for_the_rest_of_this_period();
+static void start_new_period();
+
+bool log_rate_parse_policy(const char* policy_string, log_policy_t* policy) {
+	if (policy_string == NULL) {
+		*policy = PASSTHROUGH;
+		return true;
+	}
+	if (!strcmp(policy_string, "backpressure")) {
+		*policy = BACKPRESSURE;
+		return true;
+	} else if (!strcmp(policy_string, "drop")) {
+		*policy = DROP;
+		return true;
+	} else if (!strcmp(policy_string, "ignore")) {
+		*policy = IGNORE;
+		return true;
+	} else if (!strcmp(policy_string, "passthrough")) {
+		*policy = PASSTHROUGH;
+		return true;
+	} else {
+		return false;
+	}
+}
+
+bool log_rate_parse_rate_limit(const char* rate_limit_string, size_t* rate_limit) {
+	if (rate_limit_string == NULL) {
+		rate_limit = 0;
+		return true;
+	}
+	char* endptr;
+	errno = 0;
+	long unscaled_rate_limit = strtol(rate_limit_string, &endptr, 10);
+	if (errno != 0 || unscaled_rate_limit <= 0) {
+		return false;
+	}
+	size_t scale = 1;
+	switch (*endptr) {
+	case '\0':
+		break;
+	case 'K':
+		scale = (size_t)1024;
+		break;
+	case 'M':
+		scale = (size_t)1024 * 1024;
+		break;
+	case 'G':
+		scale = (size_t)1024 * 1024 * 1024;
+		break;
+	case 'T':
+		scale = (size_t)1024 * 1024 * 1024 * 1024;
+		break;
+	default:
+		return false;
+	}
+	*rate_limit = unscaled_rate_limit * scale;
+	return true;
+}
+
+void log_rate_init(log_policy_t policy, size_t rate_limit) {
+	log_policy = policy;
+	bytes_per_period = rate_limit;
+	start_new_period();
+}
+
+bool log_rate_write_to_logs(stdpipe_t pipe, char *buf, ssize_t num_read) {
```

Comment:
Perhaps it'd be worth a comment here stating that this function's call signature must be kept the same as `write_to_logs()`?

Reply:
Also, it is not clear this is the interface we want to rate limit.  The SRE needs to be able to control the final behavior of `conmon`, after it has added all the metadata to the logs.  So it would seem we need to add the rate limit to the file writes to disk, or the journald writes.

Reply:
Yep, that's true. Stay tuned...